packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart

```diff
@@ -139,32 +142,110 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
+          final expiredLogEventEndIndex =
+              rejectedLogEventsInfo?.expiredLogEventEndIndex;
+          final tooOldLogEventEndIndex =
+              rejectedLogEventsInfo?.tooOldLogEventEndIndex;
+          final tooNewLogEventStartIndex =
+              rejectedLogEventsInfo?.tooNewLogEventStartIndex;
+
+          if (expiredLogEventEndIndex != null &&
+              expiredLogEventEndIndex >= 0 &&
+              expiredLogEventEndIndex <= events.length - 1) {
+            // delete old logs from log store
+            await _logStore
+                .deleteItems(logs.sublist(0, expiredLogEventEndIndex + 1));
+            // set logs to start from next
+            logs.removeRange(0, expiredLogEventEndIndex + 1);
+            // set events to start from next
+            events.removeRange(0, expiredLogEventEndIndex + 1);
+            continue;
+          }
+          if (tooOldLogEventEndIndex != null &&
+              tooOldLogEventEndIndex >= 0 &&
+              tooOldLogEventEndIndex <= events.length - 1) {
+            // delete old logs from log store
+            await _logStore
+                .deleteItems(logs.sublist(0, tooOldLogEventEndIndex + 1));
+            // set logs to start from next
+            logs.removeRange(0, tooOldLogEventEndIndex + 1);
+            // set events to start from next
+            events.removeRange(0, tooOldLogEventEndIndex + 1);
+            continue;
+          }
+          if (tooNewLogEventStartIndex != null &&
+              tooNewLogEventStartIndex >= 0 &&
+              tooNewLogEventStartIndex <= events.length - 1) {
+            tooNewException = _TooNewLogEventException(
+              events[tooNewLogEventStartIndex].timestamp.toInt(),
+            );
+            // set logs to end at the index
+            logs.removeRange(tooNewLogEventStartIndex, logs.length);
+            // set events to end at the index
+            events.removeRange(tooNewLogEventStartIndex, events.length);
+            continue;
+          }
+          await _logStore.deleteItems(logs);
           break;
         }
-        await _logStore.deleteItems(logs);
+        if (tooNewException != null) {
+          throw tooNewException;
+        }
       }
     }
 
     if (!_syncing) {
-      // TODO(nikahsn): disable log rotation.
       _syncing = true;
+      var nextRetry = 0;
       try {
         await startSyncing();
+      } on _TooNewLogEventException catch (e) {
+        nextRetry = e.timeInMillisecondsSinceEpoch -
+            _maxLogEventTimeInFutureInMilliseconds;
       } on Exception catch (e) {
         logger.error('Failed to sync logs to CloudWatch.', e);
-        // TODO(nikahsn): enable log rotation if the log store is full
       } finally {
+        _handleFullLogStoreAfterSync(
+          retryTimeInMillisecondsSinceEpoch: nextRetry,
+        );
         _syncing = false;
       }
     }
   }
 
+  void _handleFullLogStoreAfterSync({
+    int retryTimeInMillisecondsSinceEpoch = 0,
```

Comment:
Can you change this logic to work with `DateTime` and `Duration` objects instead of integers? This is very opaque and hard to follow